addons/website_sale_stock/models/product_product.py

```diff
@@ -34,7 +34,9 @@ def _send_availability_email(self):
         for product in self.search([('stock_notification_partner_ids', '!=', False)]):
             if product._is_sold_out():
                 continue
+            ctx_lang = self._context.get('lang')
             for partner in product.stock_notification_partner_ids:
+                self = self.with_context(lang=partner.lang if partner.lang else 'en_US')
                 body_html = self.env['ir.qweb']._render('website_sale_stock.availability_email_body', {"product": product})
                 msg = self.env["mail.message"].sudo().new(dict(body=body_html, record_name=product.name))
                 full_mail = self.env["mail.render.mixin"]._render_encapsulate(
```

Comment:
Hello @Feyensv ,

> I don't really like the concept of modifying self so many times here, it's a bit misleading.
You should use a new variable to hold the record with the modified environment.

I tested your suggestion and the issue with using a new variable to hold the record with the modified environment is that the subject of the mail uses self's context lang and thus will not get translated accordingly. 
Source Code Line: https://github.com/odoo/odoo/blob/c40327db4f46040d4a6ed0231afcdefd5be47db5/addons/website_sale_stock/models/product_product.py#L46

Self needs to be modified by passing in partner's lang before mail_values is defined for it to be translated, which is why I had thought of  modifying self. 

The original thought process was to save the original context lang of self, modify self's context lang to be the partner's lang for each partner, and at the end reassign self's lang to its original context lang.

Feel free to let me know if you know of a better implementation to resolve the subject line's translation.

> Also, I think it might be necessary to also include the language in the context of the product record, to make sure the full content of the mail is translated (including the attributes name).

This is a great idea. I overlooked the translation of the product's record including its name. I included this into my implementation.

>Lastly, i don't think specifying en_US as fallback is necessary, fallback values should already be correctly handled by the translation logic

I omitted this as you suggested.